python/cudf/cudf/core/dtypes.py

```diff
@@ -8,9 +8,333 @@
+        self.pa_type = pa.timestamp("ns")
+        self._name = "Datetime64NS"
+        self._time_unit = "ns"
+
+
+class Datetime64USDtype(Datetime):
+    def __init__(self):
+        self.pa_type = pa.timestamp("us")
+        self._name = "Datetime64US"
+        self._time_unit = "us"
+
+
+class Datetime64MSDtype(Datetime):
+    def __init__(self):
+        self.pa_type = pa.timestamp("ms")
+        self._name = "Datetime64MS"
+        self._time_unit = "ms"
+
+
+class Datetime64SDtype(Datetime):
+    def __init__(self):
+        self.pa_type = pa.timestamp("s")
+        self._name = "Datetime64S"
+        self._time_unit = "s"
+
+
+class Timedelta64NSDtype(Timedelta):
+    def __init__(self):
+        self.pa_type = pa.duration("ns")
+        self._name = "Timedelta64NS"
+        self._time_unit = "ns"
+
+
+class Timedelta64USDtype(Timedelta):
+    def __init__(self):
+        self.pa_type = pa.duration("us")
+        self._name = "Timedelta64US"
+        self._time_unit = "us"
+
+
+class Timedelta64MSDtype(Timedelta):
+    def __init__(self):
+        self.pa_type = pa.duration("ms")
+        self._name = "Timedelta64MS"
+        self._time_unit = "ms"
+
+
+class Timedelta64SDtype(Timedelta):
+    def __init__(self):
+        self.pa_type = pa.duration("s")
+        self._name = "Timedelta64S"
+        self._time_unit = "s"
+
+
+class StringDtype(Flexible):
+    def __init__(self):
+        self.pa_type = pa.string()
+        self._name = "string"
+
+
+class CUDFType(object):
+    def __init__(self, parent_dtype):
+        self.parent_dtype = parent_dtype
+
+    def __call__(self, arg):
+        return cudf.Scalar(arg, dtype=self.parent_dtype)
+
+
+def cudf_dtype_from_string(obj):
+    if obj == "category":
+        return obj
+    try:
+        np_dtype = np.dtype(obj)
+        return cudf_dtype_from_numpy(np_dtype)
+    except TypeError:
+        result = _cudf_dtype_from_string.get(obj, None)
+        if not result:
+            raise TypeError(f"Could not find a cuDF dtype matching {obj}")
+        return result
+
+
+def cudf_dtype_from_numpy(obj):
+    if obj is np.str_:
+        return StringDtype()
+    elif obj is np.number:
+        return cudf.Number
+    elif obj is np.datetime64:
+        return cudf.Datetime
+    elif obj is np.timedelta64:
+        return cudf.Timedelta
+    dtype = np.dtype(obj)
+    if dtype.type is np.str_:
+        return StringDtype()
+    result = _cudf_dtype_from_numpy.get(dtype, None)
+    if not result:
+        raise TypeError(f"Could not find a cuDF dtype matching {obj}")
+    return result
+
+
+def cudf_dtype_from_pandas(obj):
+    if isinstance(obj, pd.core.arrays.numpy_.PandasDtype):
+        try:
+            return cudf_dtype_from_numpy(obj.numpy_dtype)
+        except TypeError:
+            result = _cudf_dtype_from_pandas.get(obj, None)
+            if not result:
+                raise TypeError(f"Could not find a cuDF dtype matching {obj}")
+            return result
+
+
+def dtype(obj):
```

Comment:
What is the intended return value of this API? Looks like it's allowed to return either a `Dtype` or `None`? When is it OK to return `None`?

Reply:
good question - this might be left over. I think originally it was because we don't want to force a dtype inside certain places in column factory (`as_column`) - as we like to let pyarrow infer it and then just pass that result to the arrow related pieces of the column factory.

That said, `np.dtype(None) == np.dtype('float64')`. Ideally we could have this compatibility. The price would be statements like `dtype = cudf.dtype(dtype) if dtype is not None else None`. I think this is what I was trying to avoid.